java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java

```diff
@@ -0,0 +1,297 @@
+    scan(variables);
+    for (VariableTree methodParameter : variables) {
+      Symbol symbol = methodParameter.symbol();
+      if (isCloseableOrAutoCloseableSubtype(symbol.type())) {
+        executionState.markValueAs(symbol, new CloseableState.Ignored(methodParameter));
+      }
+    }
+  }
+
+  private static MethodInvocationMatcherCollection closeMethodInvocationMatcher() {
+    return MethodInvocationMatcherCollection.create(
+      MethodInvocationMatcher.create()
+        .typeDefinition(TypeCriteria.subtypeOf(JAVA_IO_CLOSEABLE))
+        .name(CLOSE_METHOD_NAME)
+        .withNoParameterConstraint(),
+      MethodInvocationMatcher.create()
+        .typeDefinition(TypeCriteria.subtypeOf(JAVA_LANG_AUTOCLOSEABLE))
+        .name(CLOSE_METHOD_NAME)
+        .withNoParameterConstraint());
+  }
+
+  private static boolean isCloseableOrAutoCloseableSubtype(Type type) {
+    return type.isSubtypeOf(JAVA_IO_CLOSEABLE) || type.isSubtypeOf(JAVA_LANG_AUTOCLOSEABLE);
+  }
+
+  private static boolean isIgnoredCloseableSubtype(Type type) {
+    for (String fullyQualifiedName : IGNORED_CLOSEABLE_SUBTYPES) {
+      if (type.isSubtypeOf(fullyQualifiedName)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private static boolean isSubclassOfInputStreamOrOutputStreamWithoutClose(Type type) {
+    Symbol.TypeSymbol typeSymbol = type.symbol();
+    Type superClass = typeSymbol.superClass();
+    if (superClass != null && (superClass.is("java.io.OutputStream") || superClass.is("java.io.InputStream"))) {
+      return typeSymbol.lookupSymbols(CLOSE_METHOD_NAME).isEmpty();
+    }
+    return false;
+  }
+
+  @Override
+  protected boolean isSymbolRelevant(Symbol symbol) {
+    return isCloseableOrAutoCloseableSubtype(symbol.type());
+  }
+
+  @Override
+  public void visitVariable(VariableTree tree) {
+    super.visitVariable(tree);
+    ExpressionTree initializer = tree.initializer();
+
+    // check first usage of closeables in order to manage use of same symbol
+    ignoreClosableSymbols(initializer);
+
+    Symbol symbol = tree.symbol();
+    if (isCloseableOrAutoCloseableSubtype(symbol.type())) {
+      executionState.markValueAs(symbol, getCloseableStateFromExpression(symbol, initializer));
+    }
+  }
+
+  private CloseableState getCloseableStateFromExpression(Symbol symbol, @Nullable ExpressionTree expression) {
+    if (shouldBeIgnored(symbol, expression)) {
+      return new CloseableState.Ignored(expression);
+    } else if (isNull(expression)) {
+      return new CloseableState.Null(expression);
+    } else if (expression.is(Tree.Kind.NEW_CLASS)) {
+      if (usesIgnoredCloseableAsArgument(((NewClassTree) expression).arguments())) {
+        return new CloseableState.Ignored(expression);
+      }
+      return new CloseableState.Open(expression);
+    }
+    // TODO SONARJAVA-1029 : Engine currently ignore closeables which are retrieved from method calls. Handle them as OPEN.
```

Comment:
:new: ![INFO](https://raw.githubusercontent.com/henryju/image-hosting/master/info.png)Complete the task associated to this TODO comment.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)